importer - Assertion `!isIncompleteType() && "This doesn't make sense for incomplete types

clang crashes while analyzing FFmpeg/libavformat/act.c


/ssd/ednikru/ctu/clang_github-fork/clang/tools/scan-build-py/bin/analyze-cc -I/ssd/ednikru/ctu/tests/FFmpeg -I/ssd/ednikru/ctu/tests/FFmpeg -D_ISOC99_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_POSIX_C_SOURCE=200112 -D_XOPEN_SOURCE=600 -I/ssd/ednikru/ctu/tests/FFmpeg/compat/atomics/gcc -I/usr/include/c++/4.8 -I/usr/include/x86_64-linux-gnu/c++/4.8 -I/usr/include/c++/4.8/backward -I/usr/local/include -I/usr/include/x86_64-linux-gnu -I/usr/include -DZLIB_CONST -DHAVE_AV_CONFIG_H -std=c11 -fomit-frame-pointer -pthread -g -Wdeclaration-after-statement -Wall -Wdisabled-optimization -Wpointer-arith -Wredundant-decls -Wwrite-strings -Wtype-limits -Wundef -Wmissing-prototypes -Wno-pointer-to-int-cast -Wstrict-prototypes -Wempty-body -Wno-parentheses -Wno-switch -Wno-format-zero-length -Wno-pointer-sign -O3 -fno-math-errno -fno-signed-zeros -fno-tree-vectorize -Werror=format-security -Werror=implicit-function-declaration -Werror=missing-prototypes -Werror=return-type -Werror=vla -Wformat -Wno-maybe-uninitialized -MMD -MF libavformat/act.d -MT libavformat/act.o -c -o /ssd/ednikru/ctu/tests/FFmpeg/libavformat/act.o /ssd/ednikru/ctu/tests/FFmpeg/libavformat/act.c
analyze: INFO: execute compiler: gcc -I/ssd/ednikru/ctu/tests/FFmpeg -I/ssd/ednikru/ctu/tests/FFmpeg -D_ISOC99_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_POSIX_C_SOURCE=200112 -D_XOPEN_SOURCE=600 -I/ssd/ednikru/ctu/tests/FFmpeg/compat/atomics/gcc -I/usr/include/c++/4.8 -I/usr/include/x86_64-linux-gnu/c++/4.8 -I/usr/include/c++/4.8/backward -I/usr/local/include -I/usr/include/x86_64-linux-gnu -I/usr/include -DZLIB_CONST -DHAVE_AV_CONFIG_H -std=c11 -fomit-frame-pointer -pthread -g -Wdeclaration-after-statement -Wall -Wdisabled-optimization -Wpointer-arith -Wredundant-decls -Wwrite-strings -Wtype-limits -Wundef -Wmissing-prototypes -Wno-pointer-to-int-cast -Wstrict-prototypes -Wempty-body -Wno-parentheses -Wno-switch -Wno-format-zero-length -Wno-pointer-sign -O3 -fno-math-errno -fno-signed-zeros -fno-tree-vectorize -Werror=format-security -Werror=implicit-function-declaration -Werror=missing-prototypes -Werror=return-type -Werror=vla -Wformat -Wno-maybe-uninitialized -MMD -MF libavformat/act.d -MT libavformat/act.o -c -o /ssd/ednikru/ctu/tests/FFmpeg/libavformat/act.o /ssd/ednikru/ctu/tests/FFmpeg/libavformat/act.c
analyze: INFO: Importing function _Z13av_new_packetP8AVPacketi@x86_64 from /ssd/ednikru/ctu/tests/FFmpeg/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/FFmpeg/libavcodec/avpacket.c.ast
analyze: INFO: Importing function _Z17av_buffer_reallocPP11AVBufferRefi@x86_64 from /ssd/ednikru/ctu/tests/FFmpeg/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/FFmpeg/libavutil/buffer.c.ast
analyze: INFO: Importing function _Z10av_reallocPvm@x86_64 from /ssd/ednikru/ctu/tests/FFmpeg/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/FFmpeg/libavutil/mem.c.ast
analyze: INFO: clang-4.0: /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/Type.cpp:1903: bool clang::Type::isConstantSizeType() const: Assertion `!isIncompleteType() && "This doesn't make sense for incomplete types"' failed.
analyze: INFO: #0 0x00000000037e0f7d llvm::sys::PrintStackTrace(llvm::raw_ostream&) /home/ednikru/ssd/ctu/llvm/lib/Support/Unix/Signals.inc:402:0
analyze: INFO: #1 0x00000000037e12d8 PrintStackTraceSignalHandler(void*) /home/ednikru/ssd/ctu/llvm/lib/Support/Unix/Signals.inc:466:0
analyze: INFO: #2 0x00000000037df49e llvm::sys::RunSignalHandlers() /home/ednikru/ssd/ctu/llvm/lib/Support/Signals.cpp:44:0
analyze: INFO: #3 0x00000000037e090f SignalHandler(int) /home/ednikru/ssd/ctu/llvm/lib/Support/Unix/Signals.inc:256:0
analyze: INFO: #4 0x00007fbe61ff6330 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x10330)
analyze: INFO: #5 0x00007fbe60df1c37 gsignal /build/eglibc-oGUzwX/eglibc-2.19/signal/../nptl/sysdeps/unix/sysv/linux/raise.c:56:0
analyze: INFO: #6 0x00007fbe60df5028 abort /build/eglibc-oGUzwX/eglibc-2.19/stdlib/abort.c:91:0
analyze: INFO: #7 0x00007fbe60deabf6 __assert_fail_base /build/eglibc-oGUzwX/eglibc-2.19/assert/assert.c:92:0
analyze: INFO: #8 0x00007fbe60deaca2 (/lib/x86_64-linux-gnu/libc.so.6+0x2fca2)
analyze: INFO: #9 0x0000000006542a96 clang::Type::isConstantSizeType() const /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/Type.cpp:1904:0
analyze: INFO: #10 0x000000000642a854 HandleSizeof((anonymous namespace)::EvalInfo&, clang::SourceLocation, clang::QualType, clang::CharUnits&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:2195:0
analyze: INFO: #11 0x0000000006443e05 (anonymous namespace)::IntExprEvaluator::VisitUnaryExprOrTypeTraitExpr(clang::UnaryExprOrTypeTraitExpr const*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:8439:0
analyze: INFO: #12 0x0000000006460073 clang::StmtVisitorBase<clang::make_const_ptr, (anonymous namespace)::IntExprEvaluator, bool>::Visit(clang::Stmt const*) /home/ednikru/ssd/ctu/build_ctu_master/tools/clang/include/clang/AST/StmtNodes.inc:809:0
analyze: INFO: #13 0x0000000006449e60 Evaluate(clang::APValue&, (anonymous namespace)::EvalInfo&, clang::Expr const*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:9567:0
analyze: INFO: #14 0x000000000644a790 EvaluateAsRValue((anonymous namespace)::EvalInfo&, clang::Expr const*, clang::APValue&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:9656:0
analyze: INFO: #15 0x000000000644abab clang::Expr::EvaluateAsRValue(clang::Expr::EvalResult&, clang::ASTContext const&) const /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:9711:0
analyze: INFO: #16 0x000000000644b555 clang::Expr::EvaluateKnownConstInt(clang::ASTContext const&, llvm::SmallVectorImpl<std::pair<clang::SourceLocation, clang::PartialDiagnostic> >*) const /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:9821:0
analyze: INFO: #17 0x0000000005589e85 clang::ento::ExprEngine::VisitUnaryExprOrTypeTraitExpr(clang::UnaryExprOrTypeTraitExpr const*, clang::ento::ExplodedNode*, clang::ento::ExplodedNodeSet&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngineC.cpp:840:0
analyze: INFO: #18 0x00000000055689fc clang::ento::ExprEngine::Visit(clang::Stmt const*, clang::ento::ExplodedNode*, clang::ento::ExplodedNodeSet&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:1359:0
analyze: INFO: #19 0x0000000005564aa0 clang::ento::ExprEngine::ProcessStmt(clang::CFGStmt, clang::ento::ExplodedNode*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:507:0
analyze: INFO: #20 0x0000000005563da0 clang::ento::ExprEngine::processCFGElement(clang::CFGElement, clang::ento::ExplodedNode*, unsigned int, clang::ento::NodeBuilderContext*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:355:0
analyze: INFO: #21 0x0000000005552b2d clang::ento::CoreEngine::HandlePostStmt(clang::CFGBlock const*, unsigned int, clang::ento::ExplodedNode*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:532:0
analyze: INFO: #22 0x00000000055519a0 clang::ento::CoreEngine::dispatchWorkItem(clang::ento::ExplodedNode*, clang::ProgramPoint, clang::ento::WorkListUnit const&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:279:0
analyze: INFO: #23 0x0000000005551580 clang::ento::CoreEngine::ExecuteWorkList(clang::LocationContext const*, unsigned int, llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:216:0
analyze: INFO: #24 0x0000000005034dea clang::ento::ExprEngine::ExecuteWorkList(clang::LocationContext const*, unsigned int) /home/ednikru/ssd/ctu/llvm/tools/clang/include/clang/StaticAnalyzer/Core/PathSensitive/ExprEngine.h:110:0
analyze: INFO: #25 0x0000000004feccc4 (anonymous namespace)::AnalysisConsumer::ActionExprEngine(clang::Decl*, bool, clang::ento::ExprEngine::InliningModes, llvm::DenseSet<clang::Decl const*, llvm::DenseMapInfo<clang::Decl const*> >*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:826:0
analyze: INFO: #26 0x0000000004fed0b4 (anonymous namespace)::AnalysisConsumer::RunPathSensitiveChecks(clang::Decl*, clang::ento::ExprEngine::InliningModes, llvm::DenseSet<clang::Decl const*, llvm::DenseMapInfo<clang::Decl const*> >*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:862:0
analyze: INFO: #27 0x0000000004feca8a (anonymous namespace)::AnalysisConsumer::HandleCode(clang::Decl*, unsigned int, clang::ento::ExprEngine::InliningModes, llvm::DenseSet<clang::Decl const*, llvm::DenseMapInfo<clang::Decl const*> >*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:792:0
analyze: INFO: #28 0x0000000004feb401 (anonymous namespace)::AnalysisConsumer::HandleDeclsCallGraph(unsigned int) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:575:0
analyze: INFO: #29 0x0000000004febe49 (anonymous namespace)::AnalysisConsumer::HandleTranslationUnit(clang::ASTContext&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:657:0
analyze: INFO: #30 0x0000000005628933 clang::ParseAST(clang::Sema&, bool, bool) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Parse/ParseAST.cpp:161:0
analyze: INFO: #31 0x000000000401610a clang::ASTFrontendAction::ExecuteAction() /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Frontend/FrontendAction.cpp:560:0
analyze: INFO: #32 0x0000000004015baf clang::FrontendAction::Execute() /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Frontend/FrontendAction.cpp:463:0
analyze: INFO: #33 0x0000000003fbcd73 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Frontend/CompilerInstance.cpp:954:0
analyze: INFO: #34 0x000000000413d9ee clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:249:0
analyze: INFO: #35 0x0000000001cb64c1 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /home/ednikru/ssd/ctu/llvm/tools/clang/tools/driver/cc1_main.cpp:221:0
analyze: INFO: #36 0x0000000001cac87a ExecuteCC1Tool(llvm::ArrayRef<char const*>, llvm::StringRef) /home/ednikru/ssd/ctu/llvm/tools/clang/tools/driver/driver.cpp:299:0
analyze: INFO: #37 0x0000000001cad478 main /home/ednikru/ssd/ctu/llvm/tools/clang/tools/driver/driver.cpp:380:0
analyze: INFO: #38 0x00007fbe60ddcf45 __libc_start_main /build/eglibc-oGUzwX/eglibc-2.19/csu/libc-start.c:321:0
analyze: INFO: #39 0x0000000001caa2d9 _start (/ssd/ednikru/ctu/build_ctu_master/bin/clang-4.0+0x1caa2d9)
analyze: INFO: Stack dump:
analyze: INFO: 0.	Program arguments: /ssd/ednikru/ctu/build_ctu_master/bin/clang-4.0 -cc1 -triple x86_64-unknown-linux-gnu -analyze -disable-free -main-file-name act.c -analyzer-store=region -analyzer-opt-analyze-nested-blocks -analyzer-eagerly-assume -analyzer-checker=core -analyzer-checker=apiModeling -analyzer-checker=unix -analyzer-checker=deadcode -analyzer-checker=security.insecureAPI.UncheckedReturn -analyzer-checker=security.insecureAPI.getpw -analyzer-checker=security.insecureAPI.gets -analyzer-checker=security.insecureAPI.mktemp -analyzer-checker=security.insecureAPI.mkstemp -analyzer-checker=security.insecureAPI.vfork -analyzer-checker=nullability.NullPassedToNonnull -analyzer-checker=nullability.NullReturnedFromNonnull -analyzer-output plist-multi-file -w -analyzer-config xtu-dir=/ssd/ednikru/ctu/tests/FFmpeg/.xtu -analyzer-config reanalyze-xtu-visited=true -analyzer-stats -mrelocation-model static -mthread-model posix -fno-signed-zeros -masm-verbose -mconstructor-aliases -munwind-tables -fuse-init-array -target-cpu x86-64 -momit-leaf-frame-pointer -dwarf-column-info -debugger-tuning=gdb -resource-dir /ssd/ednikru/ctu/build_ctu_master/bin/../lib/clang/4.0.0 -I /ssd/ednikru/ctu/tests/FFmpeg -I /ssd/ednikru/ctu/tests/FFmpeg -D _ISOC99_SOURCE -D _FILE_OFFSET_BITS=64 -D _LARGEFILE_SOURCE -D _POSIX_C_SOURCE=200112 -D _XOPEN_SOURCE=600 -I /ssd/ednikru/ctu/tests/FFmpeg/compat/atomics/gcc -I /usr/include/c++/4.8 -I /usr/include/x86_64-linux-gnu/c++/4.8 -I /usr/include/c++/4.8/backward -I /usr/local/include -I /usr/include/x86_64-linux-gnu -I /usr/include -D ZLIB_CONST -D HAVE_AV_CONFIG_H -internal-isystem /usr/local/include -internal-isystem /ssd/ednikru/ctu/build_ctu_master/bin/../lib/clang/4.0.0/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -O3 -Wno-pointer-to-int-cast -Wno-parentheses -Wno-switch -Wno-format-zero-length -Wno-pointer-sign -Wno-maybe-uninitialized -std=c11 -fdebug-compilation-dir /ssd/ednikru/ctu/tests/FFmpeg -ferror-limit 19 -fmessage-length 0 -pthread -fobjc-runtime=gcc -fdiagnostics-show-option -vectorize-slp -o /ssd/ednikru/ctu/tests/FFmpeg/.xtu-out/report-3yDrvm.plist -x c /ssd/ednikru/ctu/tests/FFmpeg/libavformat/act.c
analyze: INFO: 1.	<eof> parser at end of file
analyze: INFO: 2.	While analyzing stack:
analyze: INFO: 	#0 AVBufferRef *av_buffer_create(uint8_t *data, int size, void (*free)(void *, uint8_t *), void *opaque, int flags)
analyze: INFO: 	#1 int av_buffer_realloc(AVBufferRef **pbuf, int size)
analyze: INFO: 	#2 static int packet_alloc(AVBufferRef **buf, int size)
analyze: INFO: 	#3 int av_new_packet(AVPacket *pkt, int size)
analyze: INFO: 	#4 static int read_packet(AVFormatContext *s, AVPacket *pkt)
analyze: INFO: 3.	/ssd/ednikru/ctu/tests/FFmpeg/libavutil/buffer.c:35:22: Error evaluating statement
analyze: INFO: 4.	/ssd/ednikru/ctu/tests/FFmpeg/libavutil/buffer.c:35:22: Error evaluating statement

--- a/lib/AST/ASTImporter.cpp
+++ b/lib/AST/ASTImporter.cpp
@@ -2674,15 +2674,27 @@ Decl *ASTNodeImporter::VisitTypedefNameDecl(TypedefNameDecl *D, bool IsAlias) {
       if (!FoundDecls[I]->isInIdentifierNamespace(IDNS))
         continue;
       if (TypedefNameDecl *FoundTypedef =
-            dyn_cast<TypedefNameDecl>(FoundDecls[I])) {
-        if (Importer.IsStructurallyEquivalent(D->getUnderlyingType(),
-                                            FoundTypedef->getUnderlyingType()))
-          return Importer.Imported(D, FoundTypedef);
+              dyn_cast<TypedefNameDecl>(FoundDecls[I])) {
+        if (Importer.IsStructurallyEquivalent(
+                D->getUnderlyingType(), FoundTypedef->getUnderlyingType())) {
+          QualType original_ut = D->getUnderlyingType();
+          QualType found_ut = FoundTypedef->getUnderlyingType();
+          // If the found definition is incomplete
+          // but it should be complete import
+          // FIXME: maybe this check should go into
+          // IsStructurallyEquivalent() function?
+          if (!original_ut->isIncompleteType() &&
+              found_ut->isIncompleteType()) {
+            continue;
+          } else {
+            return Importer.Imported(D, FoundTypedef);
+          }
+        }
       }
-      
+
       ConflictingDecls.push_back(FoundDecls[I]);
     }
-    
+
     if (!ConflictingDecls.empty()) {
       Name = Importer.HandleNameConflict(Name, DC, IDNS,
                                          ConflictingDecls.data(), 
@@ -2691,7 +2703,6 @@ Decl *ASTNodeImporter::VisitTypedefNameDecl(TypedefNameDecl *D, bool IsAlias) {
         return nullptr;
     }
   }
-  
   // Import the underlying type of this typedef;
   QualType T = Importer.Import(D->getUnderlyingType());
   if (T.isNull())
@@ -2716,7 +2727,6 @@ Decl *ASTNodeImporter::VisitTypedefNameDecl(TypedefNameDecl *D, bool IsAlias) {
   ToTypedef->setLexicalDeclContext(LexicalDC);
   Importer.Imported(D, ToTypedef);
   LexicalDC->addDeclInternal(ToTypedef);
-  
   return ToTypedef;
 }
 

--- a/test/Analysis/Inputs/externalFnMap.txt
+++ b/test/Analysis/Inputs/externalFnMap.txt
@@ -12,3 +12,5 @@ _ZN4chns4chf3Ei@x86_64 xtu-other.cpp.ast
 _ZN5mycls3fclEi@x86_64 xtu-other.cpp.ast
 _ZN4chns4chf1Ei@x86_64 xtu-other.cpp.ast
 _ZN4myns8embed_ns4fensEi@x86_64 xtu-other.cpp.ast
+_Z7avtSizev@x86_64 xtu-other.cpp.ast
+_Z7h_chaini@x86_64 xtu-chain.cpp.ast

--- a/test/Analysis/Inputs/xtu-other.cpp
+++ b/test/Analysis/Inputs/xtu-other.cpp
@@ -65,3 +65,13 @@ int chf1(int x) {
   return chf2(x);
 }
 }
+
+//Test for a crash when importing
+//typedefs
+struct AVBuffer {
+	int a;
+};
+typedef struct AVBuffer avt;
+int avtSize(void){
+	return sizeof(avt);
+}

--- a/test/Analysis/xtu-main.cpp
+++ b/test/Analysis/xtu-main.cpp
@@ -39,6 +39,10 @@ public:
 namespace chns {
 int chf1(int x);
 }
+//test for a crash
+//when inlining typedefs
+typedef struct AVBuffer avt;
+int avtSize(void);
 
 int main() {
   clang_analyzer_eval(f(3) == 2); // expected-warning{{TRUE}}
@@ -46,13 +50,12 @@ int main() {
   clang_analyzer_eval(f(5) == 3); // expected-warning{{FALSE}}
   clang_analyzer_eval(g(4) == 6); // expected-warning{{TRUE}}
   clang_analyzer_eval(h(2) == 8); // expected-warning{{TRUE}}
-
   clang_analyzer_eval(myns::fns(2) == 9);                   // expected-warning{{TRUE}}
   clang_analyzer_eval(myns::embed_ns::fens(2) == -1);       // expected-warning{{TRUE}}
   clang_analyzer_eval(mycls().fcl(1) == 6);                 // expected-warning{{TRUE}}
   clang_analyzer_eval(mycls::fscl(1) == 7);                 // expected-warning{{TRUE}}
   clang_analyzer_eval(myns::embed_cls().fecl(1) == -6);     // expected-warning{{TRUE}}
   clang_analyzer_eval(mycls::embed_cls2().fecl2(0) == -11); // expected-warning{{TRUE}}
-
   clang_analyzer_eval(chns::chf1(4) == 12); // expected-warning{{TRUE}}
+  clang_analyzer_eval(avtSize() == 4); // expected-warning{{TRUE}}
 }

